### PR TITLE
[#965] feat(tez): support remote shuffle for tez framework

### PR DESCRIPTION
**Pull request by zhengchenyu**
### What changes were proposed in this pull request?

A method to support remote shuffle for tez framework.

### How was this patch tested?

test in real cluster and integration test.

### CODE REVIEW

**Comment by zhengchenyu** [comment]
If we want to run tez job with unffile, two way must do:

* (1) Choose RssDagMaster
 use this config:

| config |  value |
| --- | --- |
| tez.am.log.level | INFO,CLA -Dsun.nio.ch.bugLevel='' org.apache.tez.dag.app.RssDAGAppMaster avoidtrim |

* (2) Rename IO Name
Rename io name in code, no config should be configured.

**Comment by jerqi** [comment]
> If we want to run tez job with unffile, two way must do:
> 
> * (1) Choose RssDagMaster
>   use this config:
> 
> config	value
> tez.am.log.level	INFO,CLA -Dsun.nio.ch.bugLevel='' org.apache.tez.dag.app.RssDAGAppMaster avoidtrim

For this point, could we use `tez.am.launch.cmd-opts` ? Because we use `yarn.app.mapreduce.am.command-opts` in the MapReduce framework.

企业微信截图_d3c6f542-ae74-42ad-9790-1c210b0d61b3
It's better to use consistent style.

**Comment by zhengchenyu** [comment]
@jerqi 

Below is the launch command of appmaster. 

```
exec /bin/bash -c "$JAVA_HOME/bin/java  -Xmx???m -Djava.io.tmpdir=$PWD/tmp -server -Djava.net.preferIPv4Stack=true -Dhadoop.metrics.log.level=WARN -XX:+PrintGCDetails -verbose:gc -XX:+PrintGCTimeStamps -XX:+UseNUMA -XX:+UseParallelGC -Dlog4j.configuratorClass=org.apache.tez.common.TezLog4jConfigurator -Dlog4j.configuration=tez-container-log4j.properties -Dyarn.app.container.log.dir=/xxx/logs/application_1683514060868_55595/container_e3956_1683514060868_55595_01_000001 -Dtez.root.logger=INFO,CLA -Dsun.nio.ch.bugLevel='' org.apache.tez.dag.app.RssDAGAppMaster avoidtrim,CLA -Dsun.nio.ch.bugLevel='' org.apache.tez.dag.app.DAGAppMaster --session 1>/xxx/logs/application_1683514060868_55595/container_e3956_1683514060868_55595_01_000001/stdout 2>/xxx/logs/application_1683514060868_55595/container_e3956_1683514060868_55595_01_000001/stderr "
```

"-XX:+PrintGCDetails -verbose:gc -XX:+PrintGCTimeStamps -XX:+UseNUMA -XX:+UseParallelGC" is the value of tez.am.launch.cmd-opts. What a pity! tez.am.launch.cmd-opts is too far way from class name. 

If we use tez.am.launch.cmd-opts, some config(eg: tez.am.log.level) must be involved into tez.am.launch.cmd-opts.

Though it is tricky, tez.am.log.level is nearest from class name, so i use it.

**Comment by jerqi** [comment]
> @jerqi
> 
> Below is the launch command of appmaster.
> 
> ```
> exec /bin/bash -c "$JAVA_HOME/bin/java  -Xmx???m -Djava.io.tmpdir=$PWD/tmp -server -Djava.net.preferIPv4Stack=true -Dhadoop.metrics.log.level=WARN -XX:+PrintGCDetails -verbose:gc -XX:+PrintGCTimeStamps -XX:+UseNUMA -XX:+UseParallelGC -Dlog4j.configuratorClass=org.apache.tez.common.TezLog4jConfigurator -Dlog4j.configuration=tez-container-log4j.properties -Dyarn.app.container.log.dir=/xxx/logs/application_1683514060868_55595/container_e3956_1683514060868_55595_01_000001 -Dtez.root.logger=INFO,CLA -Dsun.nio.ch.bugLevel='' org.apache.tez.dag.app.RssDAGAppMaster avoidtrim,CLA -Dsun.nio.ch.bugLevel='' org.apache.tez.dag.app.DAGAppMaster --session 1>/xxx/logs/application_1683514060868_55595/container_e3956_1683514060868_55595_01_000001/stdout 2>/xxx/logs/application_1683514060868_55595/container_e3956_1683514060868_55595_01_000001/stderr "
> ```
> 
> "-XX:+PrintGCDetails -verbose:gc -XX:+PrintGCTimeStamps -XX:+UseNUMA -XX:+UseParallelGC" is the value of tez.am.launch.cmd-opts. What a pity! tez.am.launch.cmd-opts is too far way from class name.
> 
> If we use tez.am.launch.cmd-opts, some config(eg: tez.am.log.level) must be involved into tez.am.launch.cmd-opts.
> 
> Though it is tricky, tez.am.log.level is nearest from class name, so I use it.

It seems that we can use to call `setProperty` method in the main method to process later -D options after I ask the JVM colleagues in our team.

**Comment by zhengchenyu** [comment]
@jerqi 
Yes, we can use setProperty. But the key problem is that we could override other config.

For example: 
-Dtez.root.logger=INFO,CLA, here INFO is the value of config 'tez.am.log.level'. If we use tez.am.launch.cmd-opts, we must override this config. 

We know that the config will override the config which the values is placed between 'tez.am.log.level' and class name.

For MR, the value of yarn.app.mapreduce.am.command-opts is placed before app master class name, if yarn.app.mapreduce.am.profile is disable.

**Comment by zhengchenyu** [comment]
@jerqi Oh, I catch it! Thank you for your advice! I will do it!

File: client-tez/src/main/java/org/apache/tez/dag/app/RssDAGAppMaster.java [hunk 1]
```diff
@@ -228,15 +242,21 @@ public static void main(String[] args) {
...
+          String[] property = args[i].split("=");
+          if (property.length < 2) {
+            System.setProperty(property[0], "");
+          } else {
+            System.setProperty(property[0], property[1]);
+          }
+        } else if (args[i].contains("--session") || args[i].contains("-s")) {
+          sessionModeCliOption = true;
+        }
+      }
```
[the remark this answers is not in the document]
Comment (zhengchenyu) [reply client-tez/src/main/java/org/apache/tez/dag/app/RssDAGAppMaster.java:257]: That's a great catch! I just found the syslog of app master was missing. I will try to solve this problem.

**Review by jerqi — APPROVED** [review]
LGTM, thanks @zhengchenyu , merged to master. Could you add some documents like https://github.com/apache/incubator-uniffle/blob/master/README.md#deploy-mapreduce-client if you have time?

**Pull request merged**